kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +83,354 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
+   * updated field and that the nested IDs are not duplicate
+   */
+  private static void validateNestedIdConsistency(
+      Map<Long, StructField> currentFieldsById, Map<Long, StructField> updatedFieldsById) {
+    Map<Long, List<NestedIdEntry>> currentFieldIdToNestedIds =
+        columnIdToNestedIds(currentFieldsById);
+    Map<Long, List<NestedIdEntry>> updatedFieldIdToNestedIds =
+        columnIdToNestedIds(updatedFieldsById);
+    /* Validate that nested field IDs exist in the schema
+    and are consistent with the previous schemas */
+    for (Map.Entry<Long, List<NestedIdEntry>> entry : updatedFieldIdToNestedIds.entrySet()) {
+      Long fieldId = entry.getKey();
+      List<NestedIdEntry> updatedNestedIdEntries = entry.getValue();
+      Set<Long> seenNestedIds = new HashSet<>();
+      for (NestedIdEntry nestedIdEntry : updatedNestedIdEntries) {
+        // Duplicate nested ID with top level field
+        if (updatedFieldIdToNestedIds.containsKey(nestedIdEntry.id)) {
+          throw new IllegalArgumentException(
+              String.format("Duplicate field id %s", nestedIdEntry.id));
+        } else if (!seenNestedIds.add(nestedIdEntry.id)) {
+          throw new IllegalArgumentException(
+              String.format("Duplicate field id %s", nestedIdEntry.id));
+        }
+      }
+
+      List<NestedIdEntry> existingNestedIdEntries = currentFieldIdToNestedIds.get(fieldId);
+
+      /* If this is not a new column make sure
+      the existing nested IDs are consistent in the updated field */
+      if (existingNestedIdEntries != null) {
+        // Group new ID entries for this field by key and compare
+        Map<String, Long> updatedNestedKeyToColumnId = new HashMap<>();
+        for (NestedIdEntry updatedNestedIdEntry : updatedNestedIdEntries) {
+          updatedNestedKeyToColumnId.put(updatedNestedIdEntry.key, updatedNestedIdEntry.id);
+        }
+
+        for (NestedIdEntry existingNestedIdEntry : existingNestedIdEntries) {
+          Long updatedNestedColumnId = updatedNestedKeyToColumnId.get(existingNestedIdEntry.key);
+          if (updatedNestedColumnId == null
+              || !updatedNestedColumnId.equals(existingNestedIdEntry.id)) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Expected field with id %s to have nested key %s",
+                    fieldId, existingNestedIdEntry.key));
+          }
+        }
+      }
+    }
+  }
+
+  private static void validateNestedIdsExist(StructField field) {
+    if (!hasNestedColumnIds(field)) {
+      if (field.getDataType() instanceof MapType) {
+        throw new IllegalArgumentException(
+            String.format("Map field %s must have at least 2 nested field ids", field.getName()));
+      } else if (field.getDataType() instanceof ArrayType) {
+        throw new IllegalArgumentException(
+            String.format("Array field %s must have at least 1 nested field id", field.getName()));
```

Comment:
for followup: nit: we could create a utility method in `Precondition` . Seems like a common pattern.

```
void require(condition, msgTemplate, args...)
```

Reply:
also you could just have one single message for both map/array